vcpkg/concepts/registries.md

```diff
@@ -3,49 +3,404 @@ title: Registries concepts
+
+vcpkg can be instructed to consider ports available in custom registries by
+using a [`vcpkg-configuration.json`
+file](../reference/vcpkg-configuration-json.md#registries). See the [Tutorial:
+Install a dependency from a Git-based registry](../consume/git-registries.md)
+article for a tutorial on how to use custom registries in your projects.
+
+## Registry structure
+
+For vcpkg to interface with a registry, it must comform with the structure
+below:
+
+* A directory named `ports`, contains the collection of ports, with each
+  subdirectory containing a specific port matching the subdirectory name. For
+  example, the files for port `foo` must be located in `ports/foo`.
+* A directory named `versions` must contain the files that comprise the [versions
+  database](#versions-database).
+  
+### Example: registry structure
+
+```
+ports/
+  foo/
+    portfile.cmake
+    vcpkg.json
+versions/
+  f-/
+    foo.json
+  baseline.json
+```
+
+## Versions database
+
+All registries, regardless of their kind, must contain a `versions` folder at the
+root of the registry which contains the _versions database_.
+
+There are two components to the versions database:
+
+* the baseline file, 
+* the version files.
+
+The baseline file is a JSON file named `baseline.json` located at the root of
+the `versions` directory. The purpose of the baseline file is to describe the
+set of versions that are considered to be the latest for each port in the
+registry. The layout of the baseline file depends on the kind of registry.
+
+For each port in the registry, a corresponding _version file_ must exist.
+The versions file is a JSON file with the same name as its corresponding port,
+for example, a port named `foo` must have a corresponding `foo.json` version
+file.
+
+vcpkg expects the versions files to be stored at the following location
+`versions/<prefix>/<port name>.json` inside the registry, where `<prefix>` is
+the first letter of the port name followed by a hyphen. For example, the
+version file for port `foo` must be stored in `versions/f-/foo.json`.
+
+The purpose of the version file is two-fold:
+
+* List all available versions of each port
+* Point to the retrieval locations of each of version.
+
+Same as the baseline file, the layout of the version file depends on the kind of
+registry.
+
+## Baseline file
 
-There are currently two options to implement your own registries: a Git-based
-registry or a filesystem-based registry.
+All registries, regardless of their kind, must contain a file named
+`baseline.json` located at `versions/baseline.json`.
+
+The purpose of the baseline file is to describe the set of versions that are
+considered to be the latest for all ports in the registry. It is expected that
+this file is updated each time a new version of a port is added to the registry.
+
+The layout of the file is a map of named baselines. With each named baseline
+being itself a map of port names to version entries.
```

Comment:
```suggestion
The file is a JSON file composed of a single object whose properties are named baseline objects. Each baseline object's properties are port names, whose values are version entries.
```

The [JSON spec](https://www.json.org/json-en.html) calls this 'object' rather than 'map'